File: benchmark_vs_sceptre.R
# Benchmark the reference SCEPTRE R package on its own example dataset.
#
# Times a calibration check (1 CPU core, then 2 cores) on
# sceptredata::highmoi_example_data and writes the timings to
# sceptre_timing.txt for benchmark_vs_sceptre.py to compare against.
#
# Requires the sceptre + sceptredata packages (see the README section
# "Benchmark vs SCEPTRE" for a setup that installs them).
#
#   Rscript benchmark_vs_sceptre.R

suppressMessages({library(sceptre); library(sceptredata)})
data(highmoi_example_data, package = "sceptredata")
data(grna_target_data_frame_highmoi, package = "sceptredata")

N_PAIRS <- 1000   # calibration pairs to test (the matched workload)

build_object <- function() {
  so <- import_data(
    response_matrix        = highmoi_example_data$response_matrix,
    grna_matrix            = highmoi_example_data$grna_matrix,
    extra_covariates       = highmoi_example_data$extra_covariates,
    grna_target_data_frame = grna_target_data_frame_highmoi,
    moi                    = "high")
  so <- set_analysis_parameters(so)
  so <- assign_grnas(so, method = "thresholding")
  run_qc(so)
}

time_calibration <- function(parallel, n_processors) {
  so <- build_object()
  t0 <- Sys.time()
  so <- run_calibration_check(so,
                              n_calibration_pairs    = N_PAIRS,
                              calibration_group_size = 1,
                              parallel               = parallel,
                              n_processors           = n_processors)
  as.numeric(difftime(Sys.time(), t0, units = "secs"))
}

t_1core <- time_calibration(FALSE, 1)
# 2-core run; n_processors is set explicitly because detectCores() is unreliable
# inside some containers (e.g. Colab). NA if the parallel backend is unavailable.
t_2core <- tryCatch(time_calibration(TRUE, 2), error = function(e) NA_real_)

rm_ <- highmoi_example_data$response_matrix
cat(sprintf("SCEPTRE  1-core: %.1fs | 2-core: %s | %d pairs | %d genes x %d cells\n",
            t_1core, ifelse(is.na(t_2core), "NA", sprintf("%.1fs", t_2core)),
            N_PAIRS, nrow(rm_), ncol(rm_)))
writeLines(as.character(c(t_1core, t_2core, N_PAIRS, nrow(rm_), ncol(rm_))),
           "sceptre_timing.txt")

File: benchmark_vs_sceptre.py
"""
Compare gpu-sceptre against the reference SCEPTRE R package.

Run benchmark_vs_sceptre.R first (it writes sceptre_timing.txt with SCEPTRE's
wall-clock on its own example dataset), then run this script. It runs gpu-sceptre
on a matched workload — the same gene x cell dimensions and the same number of
pairs — and prints the wall-clock comparison.

    Rscript benchmark_vs_sceptre.R      # SCEPTRE side -> sceptre_timing.txt
    python3 benchmark_vs_sceptre.py     # gpu-sceptre side + comparison

Honesty notes:
  * Both timings include per-gene precompute.
  * SCEPTRE runs on its real example data; gpu-sceptre runs a matched-size,
    matched-count simulated workload. Timing depends on dimensions, not values,
    so this is a fair scale comparison; it is NOT an identical-pairs benchmark
    (SCEPTRE's negative-control pairs vs simulated pairs differ).
  * Both are timed on whatever hardware you run them on; report it.
"""
import torch
import pipeline as pipe
from demo import simulate

with open("sceptre_timing.txt") as fh:
    v = fh.read().split()
t_1core = float(v[0])
t_2core = float("nan") if v[1] == "NA" else float(v[1])
n_pairs, genes, cells = int(v[2]), int(v[3]), int(v[4])

device = "cuda" if torch.cuda.is_available() else "cpu"
n_grnas = max(2, round(n_pairs / genes))          # match SCEPTRE's pair count
Y, X, grna_to_cells, _ = simulate(n=cells, G=genes, K_nt=n_grnas, K_pos=0)
res, t_pre, t_test = pipe.run_screen(Y, X, grna_to_cells, device=device,
                                     dtype=torch.float64, B1=499, B2=1999, B3=0,
                                     gene_chunk=64)
ours = t_pre + t_test
our_pairs = genes * len(grna_to_cells)

print(f"\nDataset: {genes} genes x {cells} cells | ~{n_pairs} pairs | both include precompute")
print(f"  gpu-sceptre ({device:4s})   : {ours:7.1f} s  ({our_pairs} pairs)")
print(f"  SCEPTRE   1 CPU core   : {t_1core:7.1f} s   -> {t_1core/ours:5.1f}x slower than gpu-sceptre")
if t_2core == t_2core:  # not NaN
    eff = 100.0 * (t_1core / t_2core) / 2.0
    print(f"  SCEPTRE   2 CPU cores  : {t_2core:7.1f} s   -> {t_2core/ours:5.1f}x slower "
          f"({t_1core/t_2core:.2f}x from 2 cores = {eff:.0f}% efficient)")
    print(f"\n  At SCEPTRE's per-core rate, matching this run takes ~{t_1core/ours:.0f} cores if "
          f"scaling were perfect; the measured {eff:.0f}% efficiency implies more.")
